src/uwtools/scheduler.py

```diff
@@ -1,346 +1,364 @@
 """
-Job Scheduling.
+Support for HPC schedulers.
 """
 
 from __future__ import annotations
 
-import re
-from collections import UserDict, UserList
+from abc import ABC, abstractmethod
 from collections.abc import Mapping
+from copy import deepcopy
+from dataclasses import dataclass, fields
 from pathlib import Path
 from typing import Any, Dict, List
 
+from uwtools.exceptions import UWConfigError
 from uwtools.logging import log
 from uwtools.types import OptionalPath
-from uwtools.utils.file import writable
-from uwtools.utils.memory import Memory
 from uwtools.utils.processing import execute
 
-NONEISH = [None, "", " ", "None", "none", False]
-IGNORED_ATTRIBS = ["scheduler"]
 
-
-class RequiredAttribs:
+class JobScheduler(ABC):
     """
-    Key for required attributes.
+    An abstract class for interacting with HPC schedulers.
     """
 
-    ACCOUNT = "account"
-    QUEUE = "queue"
-    WALLTIME = "walltime"
-
+    def __init__(self, props: Dict[str, Any]):
+        self._props = {k: v for k, v in props.items() if k != "scheduler"}
+        self._validate_props()
 
-class OptionalAttribs:
-    """
-    Key for optional attributes.
-    """
-
-    CORES = "cores"
-    DEBUG = "debug"
-    EXCLUSIVE = "exclusive"
-    EXPORT = "export"
-    JOB_NAME = "jobname"
-    JOIN = "join"
-    MEMORY = "memory"
-    NODES = "nodes"
-    PARTITION = "partition"
-    PLACEMENT = "placement"
-    SHELL = "shell"
-    STDERR = "stderr"
-    STDOUT = "stdout"
-    TASKS_PER_NODE = "tasks_per_node"
-    THREADS = "threads"
-
-
-class BatchScript(UserList):
-    """
-    Represents a batch script to submit to a scheduler.
-    """
+    # Public methods
 
-    def __str__(self):
+    @property
+    def directives(self) -> List[str]:
         """
-        Returns string representation.
+        Returns resource-request scheduler directives.
         """
-        shebang = "#!/bin/bash\n"
-        return str(shebang + self.content())
+        pre, sep = self._prefix, self._directive_separator
+        ds = []
+        for key, value in self._processed_props.items():
+            if key in self._managed_directives:
+                switch = self._managed_directives[key]
+                ds.append(
+                    "%s %s" % (pre, switch(value))
+                    if callable(switch)
+                    else "%s %s%s%s" % (pre, switch, sep, value)
+                )
+            else:
+                ds.append("%s %s%s%s" % (pre, key, sep, value))
```

Comment:
The two `if` conditionals deal with the cases

1. The directive is managed by this module (recognized by a canonical short name, like `jobname` instead of ``-J`` for LSF, `-N` for PBS, `--job-name` for Slurm), and its value is a `lambda` that needs to be called with the supplied value to calculate the full directive text;
2. The directive is managed but its value can be directly used; or
3. The directive is unmanaged.

In case 3, the user will need to supply the full, scheduler-specific switch and value, like `--nice` and `100` for Slurm. This is a power-user feature, and we should consider extending support for managed directives if it seems like users would benefit. But there are *lots* of directives, and trying to add comprehensive support for obscure ones might not be worth the effort.